docs/architecture/components/index.md

```diff
@@ -299,6 +309,78 @@ In addition, for `PollingComponent`:
   methods as they are placed into the generated `main.cpp` file produced by ESPHome's code generation (codegen). It's
   important to note that **these methods will be called (and, thus, variables set) *before* the `setup()` method is called.**
 
+## Additional optional methods
+
+For components that need to handle shutdown gracefully (e.g., network connections, hardware cleanup), ESPHome provides additional lifecycle methods:
+
+### Shutdown sequence
+ESPHome has two shutdown modes:
+
+**Safe shutdown** (used for OTA updates, deep sleep, and graceful reboots):
+
+1. `on_safe_shutdown()`: Called first for critical cleanup operations
+2. `on_shutdown()`: Called to initiate shutdown (send disconnect messages, stop accepting new connections)
+3. `teardown()`: Called repeatedly to gracefully close connections and flush buffers. Returns `true` when complete or `false` if more time is needed
+4. `on_powerdown()`: Called after all teardowns complete to power down hardware
+
+**Forced reboot** (used for crashes, watchdog resets, or `App.reboot()`):
+
+1. `on_shutdown()`: Called to attempt minimal cleanup
+2. System restarts immediately (no teardown or powerdown)
+
+**Method details:**
+
+- `on_safe_shutdown()`: Only called during safe shutdowns. Used for critical operations that must happen before
+  any other shutdown procedures. Not called during forced reboots or crashes.
+- `on_shutdown()`: Always called when possible. Components should start their shutdown process here (e.g., send
+  disconnect messages, stop accepting new connections) but should NOT power down hardware or close connections yet.
+- `teardown()`: Only called during safe shutdowns. This is where connections are closed and buffers are flushed.
+  The system will call this repeatedly (with a timeout) until all components return `true`.
+- `on_powerdown()`: Only called during safe shutdowns after all components have completed their teardown. This is
+  the appropriate place to power down hardware, put chips into sleep mode, or turn off power supplies.
+
+### Implementation
+
+```cpp
+float ExampleComponent::get_setup_priority() const {
+  // Return the setup priority of this component
+  // Higher values mean this component will be set up later
+  return setup_priority::DATA;
+}
```

Comment:
Not sure if copypasta or if you intended to move this here--I'm good with it either way--but if we keep it here, we should probably move the explanatory text from the previous section down into this section, too. 😄

Reply:
I moved it because it's technically optional since it has a default.  They probably don't need to think about it.  I'll move it when I get back home.

Reply:
Thank you! 🍻